monai/apps/nuclick/transforms.py

```diff
@@ -0,0 +1,469 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
+import random
+
+import numpy as np
+
+from monai.config import KeysCollection
+from monai.transforms import MapTransform, RandomizableTransform, Transform
+from monai.utils import optional_import
+
+cv2, _ = optional_import("cv2")
+measure, _ = optional_import("skimage.measure")
+morphology, _ = optional_import("skimage.morphology")
+
+
+class FlattenLabeld(MapTransform):
+    """
+    FlattenLabeld creates labels per closed object contour (defined by a connectivity). For e.g if there are
+    12 small regions of 1's it will delineate them into 12 different label classes
+    """
+
+    def __init__(self, keys: KeysCollection):
+        super().__init__(keys)
+
+    def __call__(self, data):
+        d = dict(data)
+        for key in self.keys:
+            _, labels, _, _ = cv2.connectedComponentsWithStats(d[key], 4, cv2.CV_32S)
+            d[key] = labels.astype(np.uint8)
+        return d
+
+
+class ExtractPatchd(MapTransform):
+    """
+    Extracts a patch from the given image and label, however it is based on the centroid location.
+    The centroid location is a 2D coordinate (H, W). The extracted patch is extracted around the centroid,
+    if the centroid is towards the edge, the centroid will not be the center of the image as the patch will be
+    extracted from the edges onwards
+
+    Args:
+        keys: image, label
+        centroid_key: key where the centroid values are stored
+        patch_size: size of the extracted patch
+    """
+
+    def __init__(self, keys: KeysCollection, centroid_key: str = "centroid", patch_size: int = 128):
+        super().__init__(keys)
+        self.centroid_key = centroid_key
+        self.patch_size = patch_size
+
+    def __call__(self, data):
+        d = dict(data)
+
+        centroid = d[self.centroid_key]  # create mask based on centroid (select nuclei based on centroid)
+        roi_size = (self.patch_size, self.patch_size)
+
+        for key in self.keys:
+            img = d[key]
+            x_start, x_end, y_start, y_end = self.bbox(self.patch_size, centroid, img.shape[-2:])
+            cropped = img[:, x_start:x_end, y_start:y_end]
+            d[key] = self.pad_to_shape(cropped, roi_size)
+        return d
+
+    def bbox(self, patch_size, centroid, size):
+        x, y = centroid
+        m, n = size
+
+        x_start = int(max(x - patch_size / 2, 0))
+        y_start = int(max(y - patch_size / 2, 0))
+        x_end = x_start + patch_size
+        y_end = y_start + patch_size
+        if x_end > m:
+            x_end = m
+            x_start = m - patch_size
+        if y_end > n:
+            y_end = n
+            y_start = n - patch_size
+        return x_start, x_end, y_start, y_end
+
+    def pad_to_shape(self, img, shape):
```

Comment:
`SpatialPad` could be used instead here perhaps?

Reply:
That is good feedback, change has been incorporated